Add AsusWRT Config Flow and ScannerEntity

ollo69 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->

Replacing legacy DeviceScanner with ScannerEntity

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This PR introduce config flow for AsusWRT integration and replace legacy DeviceScanner with ScannerEntity.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

ollo69 committed:
Merge pull request #1 from home-assistant/dev

Align to master

ollo69 committed:
Merge pull request #2 from home-assistant/dev

Align to master

ollo69 committed:
Merge pull request #3 from home-assistant/dev

Align to master

ollo69 committed:
Merge pull request #4 from home-assistant/dev

Align to master

ollo69 committed:
Merge pull request #5 from home-assistant/dev

Align from master

ollo69 committed:
Merge pull request #6 from home-assistant/dev

Align from master

ollo69 committed:
Merge pull request #7 from home-assistant/dev

Align from master

ollo69 committed:
Merge pull request #8 from home-assistant/dev

Align from master

ollo69 committed:
Merge pull request #9 from home-assistant/dev

Align from master

ollo69 committed:
Merge pull request #10 from home-assistant/dev

Align from master

ollo69 committed:
Merge pull request #11 from home-assistant/dev

Align from master

ollo69 committed:
Merge pull request #12 from home-assistant/dev

Align from master

ollo69 committed:
Merge pull request #13 from home-assistant/dev

Align from master

ollo69 committed:
Merge pull request #14 from home-assistant/dev

Align from master

ollo69 committed:
Merge pull request #15 from home-assistant/dev

Align from master

ollo69 committed:
Merge pull request #16 from home-assistant/dev

Align from master

ollo69 committed:
Merge pull request #17 from home-assistant/dev

Align from master

ollo69 committed:
Merge pull request #18 from home-assistant/dev

Align from master

ollo69 committed:
Merge pull request #19 from home-assistant/dev

Align from master

ollo69 committed:
Merge pull request #20 from home-assistant/dev

Merge from Master

ollo69 committed:
Merge pull request #21 from home-assistant/dev

Align from master

ollo69 committed:
Merge pull request #22 from home-assistant/dev

Align from master

ollo69 committed:
AsusWrt Config Flow and ScannerEntity

ollo69 committed:
AsusWrt Config Flow and ScannerEntity

ollo69 committed:
Fix flake8, isort and plynt

ollo69 committed:
Fix flake8, isort and plynt

ollo69 committed:
Fix isort

ollo69 committed:
Fix isort

ollo69 committed:
Force run checks

ollo69 committed:
Force run checks

ollo69 committed:
Force checks run

ollo69 committed:
Force checks run

ollo69 committed:
Implemented tests and minor changes

ollo69 committed:
Implemented tests and minor changes

ollo69 committed:
Fix flake8 and isort

ollo69 committed:
Fix flake8 and isort

ollo69 committed:
Force run checks

ollo69 committed:
Force run checks

ollo69 committed:
Fixed tests

ollo69 committed:
Fixed tests

ollo69 committed:
Add sensor tests and fix coverage

ollo69 committed:
Add sensor tests and fix coverage

ollo69 committed:
Fix isort

ollo69 committed:
Fix isort

ollo69 committed:
Added check for valid hostname and minor changes

ollo69 committed:
Added check for valid hostname and minor changes

ollo69 committed:
Improved last activity attribute formating

ollo69 committed:
Improved last activity attribute formating

ollo69 committed:
Fix pylint

ollo69 committed:
Fix pylint

ollo69 committed:
Fix wrong call

ollo69 committed:
Fix wrong call

ollo69 committed:
Last refine

ollo69 committed:
Last refine

ollo69 committed:
Merge pull request #23 from home-assistant/dev

Align from master

ollo69 committed:
Rebase and added Temperature sensors

ollo69 committed:
Rebase and added Temperature sensors

ollo69 commented:
I just wonder why @kennedyshead wasn't automatically involved in this PR?

kennedyshead commented:
Thank you @ollo69 (I wonder that myself ;) ).

There are some functions used from aioasuswrt that are not tested so far, and therefor should be viewed as alpha, the functions not tested are: `async_get_nvram`, `async_get_bytes_total`, `async_get_current_transfer_rates` and `async_add_dns_record`

ollo69 commented:
Hi @kennedyshead, thank for your answer.

Here my comment about:
- `async_get_bytes_total`, `async_get_current_transfer_rates` are already used in current integration sensor implemetation.
- `async_get_nvram` is only used once during initialization and fails is catch, so that we will just have a couple of missing device information (not critical)
- `async_add_dns_record` is not used at all.

But I introduced use of method `async_get_temperature` to provide temperature sensor.

All this sensors looks very usefull and I would not remove sensor that was available in the previous version. But may be I could add an integration option to enable temperature sensors that are total new feature? Or I should just not include for now?

Thoughts?

kennedyshead commented:
I love the temperature sensor, we should use all the sensors that we can offer imo.
I think that you should split the PR though, one for configflow and one for the sensor (just remove the sensor in this one and add a new after this is merged).
Im pretty sure you should break out the "replace legacy DeviceScanner with ScannerEntity." to a 3:rd one as well if possible.

kennedyshead commented:
And about sensors, they should be added in the same manner as other sensors are added, dunno if they are optional for most parts or added by default

ollo69 commented:
> Im pretty sure you should break out the "replace legacy DeviceScanner with ScannerEntity." to a 3:rd one as well if possible.

I don't think that this is right approach in HA. DeviceScanner is legacy interface, ScannerEntity should come together or we will have a device in Integration page with 0 device a I think that does not make sense. But anyway we can wait for reviewer opinion.

ollo69 commented:
> I think that you should split the PR though, one for configflow and one for the sensor (just remove the sensor in this one and add a new after this is merged).

This means that sensor will remain configurable only using `configuration.yaml`. Again I don't know what HA team think about this. I can remove this part from PR, but what is the advantage?

ollo69 committed:
Merge pull request #24 from home-assistant/dev

Align from master

ollo69 committed:
Rebased and sensors change

- Set innitial sensor state to disabòed
- Avoid calling sensor API methods for disabled sensor

ollo69 committed:
Merge branch 'asuswrt-config-flow' of https://github.com/ollo69/core into asuswrt-config-flow

ollo69 committed:
Fix codespell and pylint

ollo69 commented:
Content of this PR is moved to #44759
